ocp_utilities/data_collector.py

```diff
@@ -0,0 +1,165 @@
+                file_name=f"{pod_name}.log", content=pod.log(), base_directory=directory
+            )
+            write_to_file(
+                file_name=f"{pod_name}.yaml",
+                content=pod.instance.to_str(),
+                base_directory=directory,
+            )
+
+
+def collect_data(directory, resource_object, collect_pod_logs):
+    LOGGER.info(f"Collecting instance data for {resource_object.name}")
+
+    if resource_object.kind == ProjectRequest.kind:
+        resource_object = Namespace(name=resource_object.name)
+
+    write_to_file(
+        file_name=f"{resource_object.name}.yaml",
+        content=resource_object.instance.to_str(),
+        base_directory=directory,
+    )
+    if resource_object.kind == VirtualMachine.kind and resource_object.vmi:
+        collect_vmi_data(vmi=resource_object.vmi, directory=directory)
+
+    if resource_object.kind == DataVolume.kind:
+        collect_data_volume_data(dyn_client=resource_object.client, directory=directory)
+
+    if resource_object.kind == Pod.kind and collect_pod_logs:
+        container_name = [
+            container["name"]
+            for container in resource_object.instance.status.containerStatuses
+        ][0]
+        write_to_file(
+            file_name=f"{resource_object.name}.log",
+            content=resource_object.log(**{"container": container_name}),
+            extra_dir_name=resource_object.kind,
+            base_directory=directory,
+        )
+
+
+def prepare_test_data_dir(item, prefix, logs_path=None):
+    test_cls_name = item.cls.__name__ if item.cls else ""
+    test_dir_log = os.path.join(
+        logs_path or py_config["data_collector"]["data_collector_base_directory"],
+        item.fspath.dirname.split("/tests/")[-1],
+        item.fspath.basename.partition(".py")[0],
+        test_cls_name,
+        item.name,
+        prefix,
+    )
+    os.makedirs(test_dir_log, exist_ok=True)
+    return test_dir_log
+
+
+def collect_resources_yaml_instance(resources_to_collect, namespace_name=None):
+    get_kwargs = {"dyn_client": infra.get_admin_client()}
+    for _resources in resources_to_collect:
+        if _resources == Service:
+            get_kwargs["namespace"] = namespace_name
+
+        for resource_obj in _resources.get(**get_kwargs):
+            try:
+                write_to_file(
+                    file_name=f"{resource_obj.name}.yaml",
+                    content=resource_obj.instance.to_str(),
+                    extra_dir_name=resource_obj.kind,
+                )
+            except Exception as exp:
+                LOGGER.warning(
+                    f"Failed to collect resource: {resource_obj.kind} {resource_obj.name} {exp}"
+                )
+
+
+def collect_pods_data(pods, pod_list=None, pods_prefix=None):
+    for pod in pods:
+        kwargs = {}
+        for pod_prefix in pod_list:
+            if pod.name.startswith(pod_prefix):
+                if pod_prefix == "virt-launcher":
```

Comment:
May be we should do what you are doing above and get the container information from the pod itself, instead of hard coding?

Reply:
updating to a generic solution